### ISSUE TITLE
some confusions.

### ISSUE THREAD

**Issue by sezan92** [philipperemy/keras-attention-mechanism#18]
![attention_luong](https://user-images.githubusercontent.com/11025093/39625088-bbd45082-4fbd-11e8-85c1-5895f1824bed.png)
Hello , Thanks for an easy code to read. But i have some confusions.

1) your attention functions takes the hidden state of input i.e lstm outputs from encoders and then does all the processes then. but according to what I have read , it must form some kind of function with the hidden state of the target , like in the given picture . Why haven't you did that ? otherwise you are just making an lstm function manually.

2) Why have you used permute layers before softmax layer ?
3) why have you averaged the outputs of softmax layer ?

**Comment by dsleo** [comment]
Hello,

Thank you for a really cool repo ! Just like @sezan92 I'm not sure of certain things regarding your attention for LSTM. 
Specifically, it's not clear why you would permute `time_steps` and `input_dim` in [here](https://github.com/philipperemy/keras-attention-mechanism/blob/master/attention_lstm.py#L18) ?

Looking at [this feed forward attention](http://colinraffel.com/publications/iclr2016feed.pdf) for a reference, we would want to apply a dense layer on top of the matrix of shape `(timestep,input_dim)` so that we  actually multiply each hidden state `h_i` by the weight matrix. 

Am I missing something here?

**Comment by XiaofengZHOU** [comment]
@dsleo 
in fact, the operation that we do for attention is to apply a tanh function for the hidden state of LSTM(or rnn). In keras, lstm will return the output, and the output we get is the result of tanh(hidden state). 
 
![attention](https://user-images.githubusercontent.com/22477810/42072958-3ac064b8-7b96-11e8-89ca-86a91f599299.png)

**Comment by zfb0014** [comment]
Hello,
Thanks for the code.
but I agree with @sezan92 . I think that it should use the hidden state to calculate the attention.
Thanks in advance for responses.

**Comment by philipperemy** [comment]
Thanks for your interest! I updated the code to reflect some changes based on this thread: https://github.com/philipperemy/keras-attention-mechanism/issues/14. Let me know if you still have some concerns :)
